src/sql/src/plan/lowering.rs

```diff
@@ -112,6 +112,21 @@ impl ColumnMap {
     }
 }
 
+/// Map with the CTEs currently in scope.
+type CteMap = HashMap<expr::LocalId, CteDesc>;
+
+/// Information about needed when finding a reference to a CTE in scope.
+struct CteDesc {
+    /// The new ID assigned to the lowered vesion of the CTE, which may not match
+    /// the ID of the input CTE.
+    new_id: expr::LocalId,
+    /// The relation type of the CTE including the columns from the outer
+    /// context at the beginning.
+    relation_type: RelationType,
+    /// The outer relation the CTE was applied to.
+    outer_relation: expr::MirRelationExpr,
```

Comment:
Is it actually necessary to store the full `MirRelationExpr`? It seems you can just store the arity of the outer relation?

Reply:
Good point. The arity is already known by storing the `relation_type`.

Reply:
It turns out keeping and comparing the outer relation, instead of the just the arity, is actually needed.

Check these two different decorrelated plans for the same input query before and after removing the `outer_relation`:

![cte-arity-before](https://user-images.githubusercontent.com/1282736/147571852-881e0c0e-4078-47ea-936d-6dfd2bf5a07f.png)
![cte-arity-after](https://user-images.githubusercontent.com/1282736/147571875-a0ce9933-7c9a-4296-b557-92f6a7feffea.png)

The `after` plan is missing a join between `l7` and `l4` (labeled as `l8`). Instead of applying `l4` (the CTE) to `l7`, we are just returning `l4` since `l7` has the same cardinality as the outer relation `l4` was applied to (`l2`), but we are missing all the operations performed from `l2` to `l7`.

Reply:
Ok. In that case, add a comment that the intermediate contexts can change the outer relation but not add any new columns.

Reply:
Done. Added an example in the comment. Same comment can be seen in the minimal test that flags the need of keeping the `outer_relation` around.